compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.cpp

```diff
@@ -0,0 +1,387 @@
+
+#include <vector>
+
+namespace
+{
+
+class Shape
+{
+  friend class SECursor;
+
+private:
+  struct ShapeElement
+  {
+    uint32_t value;
+    std::vector<char> tags;
+  };
+
+public:
+  void add_elem(int v)
+  {
+    ShapeElement se;
+    se.value = v;
+    elements.push_back(ShapeElement(se));
+  }
+
+  void add_elem_with_tag(int v, char tag)
+  {
+    ShapeElement se;
+    se.value = v;
+    se.tags.push_back(tag);
+    elements.push_back(ShapeElement(se));
+  }
+
+  void add_elem(ShapeElement e) { elements.push_back(e); }
+
+  std::vector<ShapeElement> elements;
+};
+
+class SECursor
+{
+private:
+  uint32_t _start_idx; // [start_idx, _end_idx) points Shape.elements range
+  uint32_t _end_idx;
+  uint32_t _product; // product of [se[_start_idx], se[_end_idx -1]]
+
+  std::vector<Shape::ShapeElement> &se;
+
+public:
+  SECursor(Shape &s) : _start_idx(0), _end_idx(1), se(s.elements)
+  {
+    _product = s.elements.at(0).value;
+  }
+
+  void move_next_range()
+  {
+    _start_idx = _end_idx;
+    _end_idx = _end_idx + 1;
+
+    if (_start_idx >= se.size())
+      return;
+
+    _product = se.at(_start_idx).value;
+  }
+
+  void expand_range()
+  {
+    if (_end_idx >= se.size())
+      return;
+
+    _product *= se.at(_end_idx).value;
+    _end_idx++;
+  }
+
+  // get tags all tags from pointed elements
+  std::vector<char> get_tags()
+  {
+    std::vector<char> tags;
+    for (uint32_t i = _start_idx; i < _end_idx; i++)
+    {
+      const auto &indexed_tag = se.at(i).tags;
+      tags.insert(std::end(tags), std::begin(indexed_tag), std::end(indexed_tag));
+    }
+    return tags;
+  }
+
+  // append tags at the end of each pointed elements' tag
+  void append_tags(std::vector<char> tags)
+  {
+    for (uint32_t i = _start_idx; i < _end_idx; i++)
+    {
+      auto &indexed_se = se.at(i);
+
+      if (indexed_se.value == 1)
+        continue;
+
+      indexed_se.tags.insert(std::end(indexed_se.tags), std::begin(tags), std::end(tags));
+    }
+  }
+
+  void append_incremental_tags(const char tag)
+  {
+    char incremental_tag = tag;
+    for (uint32_t i = _start_idx; i < _end_idx; i++)
+    {
+      auto &indexed_se = se.at(i);
+
+      if (indexed_se.value == 1)
+        continue;
+
+      indexed_se.tags.push_back(incremental_tag++);
+    }
+  }
+
+  bool is_end() const { return se.size() <= _start_idx; }
+
+  int size() const { return _end_idx - _start_idx; }
+
+  uint32_t product() const { return _product; }
+};
+
+class PatternSimulator
+{
+public:
+  PatternSimulator(const luci::CircleTranspose *f_tr, const luci::CircleReshape *m_rs,
+                   const luci::CircleTranspose *b_tr)
+    : _front_transpose(f_tr), _mid_reshape(m_rs), _back_transpose(b_tr), main_tag('A'), sub_tag('a')
+  {
+    assert(main_tag < sub_tag);
+  }
+
+  void simulate();
+
+  bool areTransposesUnnecessary() const;
+
+  luci::CircleReshape *createAlternative(loco::Graph *graph);
+
+private:
+  void _init_shape(const luci::CircleNode *in_tensor);
```

Comment:
Q) ONE compiler uses `_` prefix in variables not methods. Do you have any particular reason using this?